triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -1190,6 +1373,97 @@ object Runner {
     }
   }
 
+  private def isSubmissionFailure(svalue: SValue): Boolean = {
+    // svalue: Message
+    val result = for {
+      values <- svalue.expect(
+        "SVariant",
+        {
+          case SVariant(_, "MCompletion", _, SRecord(_, _, values)) if values.size() >= 1 =>
+            values.get(1)
+        },
+      )
+      _ <- values.expect("SVariant", { case SVariant(_, "Failed", _, _) => true })
+    } yield ()
+
+    result.isRight
+  }
+
+  private def isSubmissionSuccess(svalue: SValue): Boolean = {
+    // svalue: Message
+    val result = for {
+      values <- svalue.expect(
+        "SVariant",
+        {
+          case SVariant(_, "MCompletion", _, SRecord(_, _, values)) if values.size() >= 1 =>
+            values.get(1)
+        },
+      )
+      _ <- values.expect("SVariant", { case SVariant(_, "Succeeded", _, _) => true })
+    } yield ()
+
+    result.isRight
+  }
+
+  private def isTransaction(svalue: SValue): Boolean = {
+    // svalue: Message
+    svalue.expect("SVariant", { case SVariant(_, "MTransaction", _, _) => true }).isRight
+  }
+
+  private def isCreateEvent(svalue: SValue): Boolean = {
+    // svalue: Event
+    svalue.expect("SVariant", { case SVariant(_, "CreatedEvent", _, _) => true }).isRight
+  }
+
+  private def numberOfCreateEvents(svalue: SValue): Int = {
+    // svalue: Message
+    val result = for {
+      values <- svalue.expect(
+        "SVariant",
+        {
+          case SVariant(_, "MTransaction", _, SRecord(_, _, values)) if values.size() >= 2 =>
+            values.get(2)
+        },
+      )
+      events <- values.expect("SList", { case SList(events) => events.toImmArray })
+    } yield events.filter(isCreateEvent).length
+
+    result.getOrElse(0)
+  }
+
+  private def isArchiveEvent(svalue: SValue): Boolean = {
+    // svalue: Event
+    svalue.expect("SVariant", { case SVariant(_, "ArchivedEvent", _, _) => true }).isRight
+  }
+
+  private def numberOfArchiveEvents(svalue: SValue): Int = {
+    // svalue: Message
+    val result = for {
+      values <- svalue.expect(
+        "SVariant",
+        {
+          case SVariant(_, "MTransaction", _, SRecord(_, _, values)) if values.size() >= 2 =>
+            values.get(2)
+        },
+      )
+      events <- values.expect("SList", { case SList(events) => events.toImmArray })
+    } yield events.filter(isArchiveEvent).length
+
+    result.getOrElse(0)
```

Comment:
As `Message` is a variant type, we select the non-type safe approach of `getOrElse` rather than `orConverterException` (for simplicity).